### PR TITLE
Add custom metrics and metric template

### PR DESCRIPTION
**Pull request by dkistner**
**What this PR does / why we need it**:
This PR introduce some first metrics for Shoot cluster customisations.
Beside that it introduce a metric template approach which allow to describe metrics in a more declarative way.

**Release note**:

```improvement operator
Gardener metrics exporter now exposes metrics about Shoot cluster customisations.
```

cc @dguendisch

### CODE REVIEW

File: pkg/metrics/shoot_customization.go [hunk 1]
```diff
@@ -0,0 +1,643 @@
...
+						}
+					}
+				}
+			}
+			values, labels := mapLabelAndValues(&featureGateCounters)
+			return values, labels, nil
+		},
+	},
+	{
+		Name:   fmt.Sprintf("%s_apiserver_admisionplugins_total", metricShootsCustomPrefix),
```
Comment (wyb1) [inline pkg/metrics/shoot_customization.go]:
```suggestion
		Name:   fmt.Sprintf("%s_apiserver_admissionplugins_total", metricShootsCustomPrefix),
```

File: pkg/metrics/shoot_customization.go [hunk 2]
```diff
@@ -0,0 +1,643 @@
...
+		Help:   "Count of Shoots with enabled kube apiserver admission plugins.",
+		Labels: []string{"admisionplugin"},
```
Comment (wyb1) [inline pkg/metrics/shoot_customization.go]:
```suggestion
		Labels: []string{"admissionplugin"},
```

File: pkg/metrics/shoot_customization.go [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+			shoots, ok := obj.([]*gardenv1alpha1.Shoot)
+			if !ok {
+				return nil, nil, utils.NewTypeConversionError()
+			}
+
+			var admisionPluginCounters = map[string]float64{}
+			for _, s := range shoots {
+				if s.Spec.Kubernetes.KubeAPIServer != nil {
+					for _, admissionPlugin := range s.Spec.Kubernetes.KubeAPIServer.AdmissionPlugins {
+						admisionPluginCounters[admissionPlugin.Name]++
```
Comment (wyb1) [inline pkg/metrics/shoot_customization.go]:
```suggestion
						admissionPluginCounters[admissionPlugin.Name]++
```

File: pkg/metrics/shoot_customization.go [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+					counter[0]++
+				}
+			}
+			return &counter, &[][]string{}, nil
+		},
+	},
+	{
+		Name:   fmt.Sprintf("%s_apiserver_featuregates_total", metricShootsCustomPrefix),
+		Help:   "Count of Shoots with enabled kube apiserver feature gates.",
+		Labels: []string{"featuregate"},
```
Comment (wyb1) [inline pkg/metrics/shoot_customization.go]: I do not see any metrics with the label `featuregate`. Is this expected behavior?

Comment (dkistner) [reply pkg/metrics/shoot_customization.go]: Maybe you tested it with a Gardener which does not manage Shoots with apiserver feature gate configuration.
```
$ curl localhost:2718/metrics | grep feature
# HELP garden_shoots_custom_apiserver_featuregates_total Count of Shoots with enabled kube apiserver feature gates.
# TYPE garden_shoots_custom_apiserver_featuregates_total gauge
garden_shoots_custom_apiserver_featuregates_total{featuregate="CustomResourceWebhookConversion"} 53
garden_shoots_custom_apiserver_featuregates_total{featuregate="ExpandInUsePersistentVolumes"} 1
garden_shoots_custom_apiserver_featuregates_total{featuregate="TTLAfterFinished"} 13
garden_shoots_custom_apiserver_featuregates_total{featuregate="VolumeSnapshotDataSource"} 1
```

**Comment by dkistner** [comment]
I have updated the PR.

**Comment by dkistner** [comment]
I have added also a dashboard for the Shoot custom metrics to this PR.

File: pkg/metrics/shoot_customization.go (hunk 4 above)
Comment (wyb1) [reply pkg/metrics/shoot_customization.go]: Ok I see the metrics now. Should we initialize the metrics somehow? If there are no featuregates set then `garden_shoots_custom_apiserver_featuregates_total` will be missing.

Comment (wyb1) [reply pkg/metrics/shoot_customization.go]: Since we need the featuregate label I don't think we can easily initialize this metric.

**Pull request merged**